indigo_api/data_migrations/__init__.py

```diff
@@ -94,3 +96,33 @@ def update_namespace(self, xml):
             1)
 
         return xml
+
+
+class UnnumberedParagraphsToHcontainer(AKNMigration):
+    """ Update all instances un-numbered paragraphs to hcontainers. Slaw generates these when
+    a block of text is encountered in a hierarchical element.
+
+    This ALSO changes the id of the element to match AKN 3 styles, but using the id attribute (not eId)
+
+    <paragraph id="section-1.paragraph0">
+      <content>
+        <p>text</p>
+      </content>
+    </paragraph>
+
+    becomes
+
+    <hcontainer id="section-1.hcontainer_0">
+      <content>
+        <p>text</p>
+      </content>
+    </hcontainer>
+    """
+    def migrate_act(self, act):
+        for para in act.root.xpath('//a:paragraph[not(a:num)]', namespaces={'a': act.namespace}):
+            para.tag = f'{{{act.namespace}}}hcontainer'
+            # new id is based on the number of preceding hcontainer siblings
+            num = len(para.xpath('preceding-sibling::a:hcontainer', namespaces={'a': act.namespace})) + 1
```

Comment:
I don't think this does what we want it to; it numbers hcontainers continuously throughout all Schedules (I think). Going to add a test to this to show what I mean.

Reply:
`preceding-sibling` looks only at siblings -- ie they have to have the same parent. So it won't count across schedules.